reference/cd-codefresh.md

```diff
@@ -0,0 +1,120 @@
+build.
+
+The first is `PULUMI_ACCESS_TOKEN`, which is required to authenticate with pulumi.com in order to
+perform the preview or update. You can create a new Pulumi access token specifically for your
+CI/CD job on your [Pulumi Account page](https://app.pulumi.com/account/tokens).
+
+You can either add the token on the pipeline itself as a variable, or store it globally using [Codefresh shared configuration](https://codefresh.io/docs/docs/configure-ci-cd-pipeline/shared-configuration/).
+
+![Pulumi token in Codefresh](/images/reference/codefresh/codefresh-pulumi-token.png)
+
+Next, you will also need to set environment variables specific to your cloud resource provider.
+For example, if your stack is managing resources on AWS, `AWS_ACCESS_KEY_ID` and
+`AWS_SECRET_ACCESS_KEY`.
+
+
+## Codefresh CI/CD pipeline with Pulumi
+
+In your pipeline you need at least two steps.
+
+1. A step that downloads all Pulumi dependencies according to your programming language
+1. A step that performs the actual deployment 
+
+In all cases you use a [Codefresh freestyle step](https://codefresh.io/docs/docs/codefresh-yaml/steps/freestyle/) with the Pulumi Docker image.
+
+```yaml
+  RunMyPulumiStep:
+    title: Running Pulumi inside Codefresh
+    image: pulumi/pulumi
+    commands: 
+      - pulumi login
+      - pulumi stack 
+```
+
+Pulumi will search for a login token named `PULUMI_ACCESS_TOKEN` so assuming that you have setup this variable in the pipeline, Pulumi will not ask for interactive login.
+
+For Kubernetes based deployments, there is no other setup needed. Codefresh is setting up automatically a `kubeconfig` in all freestyle steps which is the same mechanism used by `kubectl` and Pulumi to access Kubernetes clusters. You only need to select your cluster if you have more than one, using a `kubectl config use-context` command prior to running `pulumi`.
+
+For other non-Kubernetes deployments, you need to add additional environment variables in your pipeline that will be used by Pulumi for accessing your cloud provider.
+
+Here is a full example:
+
+ `codefresh.yml`
+{% highlight yaml %}
+{% raw %}
+version: '1.0'
+stages:
+  - prepare
+  - build
+  - deploy
+steps:
+  main_clone:
+    title: Cloning main repository...
+    type: git-clone
+    repo: '${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}'
+    revision: '${{CF_REVISION}}'
+    stage: prepare
+    git: github-1
+  BuildProject:
+    title: Build project
+    stage: build
+    image: pulumi/pulumi
+    commands:
+      - yarn install
+  SelectMyCluster:
+    title: Select K8s cluster
+    stage: deploy
+    image: codefresh/kubectl:1.13.3
+    commands: 
+      - kubectl config get-contexts
+      - kubectl config use-context "kostis-demo@FirstKubernetes"
+  RunPulumi:
+    title: Deploying
+    stage: deploy
+    image: pulumi/pulumi
+    commands: 
+      - pulumi login
+      - pulumi stack select dev 
+      - pulumi stack 
```

Comment:
Curious what the intent is here to call the `pulumi stack` command? Once you have selected the stack with `pulumi stack select <stack-name>`, this shouldn't be necessary.

Reply:
Getting better CI/CD logs with more information.

Reply:
Got it.